src/blog/better-class-features.md

```diff
@@ -0,0 +1,332 @@
+
+Now when V8 encounters this class:
+
+```js
+class A {
+  #a = 0;
+  b = this.#a;
+}
+```
+
+It generates the following bytecode for the `b = this.#a` initializer:
+
+```cpp
+// Load the private name symbol for `#a`
+LdaImmutableCurrentContextSlot [2]
+
+// Load the value of the property keyed by `#a` from the instance into r2
+// Note: LdaKeyedProperty is renamed to GetKeyedProperty in the refactoring
+GetKeyedProperty <this>, [2]
+
+// Use the DefineKeyedOwnProperty bytecode to store the property keyed
+// by `#a` as the value of the property keyed by `b`, that is, `b = this.#a;`
+DefineNamedOwnProperty <this>, [0], [4]
+```
+
+The original `DefineNamedOwnIC` machinery could not be simply plugged into the handling of the named public class fields, as it needed a few patches to deal with an edge case of the class fields. Previously it expected the target being initialized to be an object created by V8, which was always true for object literals, but the class fields can be initialized on user-defined objects when the class extends a base class whose constructor overrides the target:
+
+```js
+class A {
+  constructor() {
+    return new Proxy(
+      { a: 1 },
+      {
+        defineProperty(object, key, desc) {
+          console.log('object:', object);
+          console.log('key:', key);
+          console.log('desc:', desc);
+          return true;
+        }
+      });
+  }
+}
+
+class B extends A {
+  a = 2;
+  #b = 3;  // Not observable.
+}
+
+// object: { a: 1 },
+// key: 'a',
+// desc: {value: 2, writable: true, enumerable: true, configurable: true}
+new B();
+```
+
+To deal with these targets, we patched the IC to fall back to the runtime when it sees that the object being initialized is a proxy, if the field being defined already exists on the object, or if the object just has a hidden class that the IC has not seen before. It is still possible to optimize the edge cases if they become common enough, but so far it seems better to trade the performance of them for simplicity of the implementation.
+
+## Optimizing private methods
+
+### The implementation of private methods
+
+In [the specification](https://tc39.es/ecma262/#sec-privatemethodoraccessoradd), the private methods are described as if they are installed on the instances but not on the class. In order to save memory, however, V8's implementation stores the private methods along with a private brand symbol in a context associated with the class. When the constructor is invoked, V8 only stores a reference to that context in the instance, with the private brand symbol as the key.
+
+![Evaluation and instantiation of classes with private methods](/_img/better-class-features/class-evaluation-and-instantiation.svg)
```

Comment:
Great graphic! I wonder if adding another instance would make it even clearer that the context is shared by multiple instances.